src/game.cpp

```diff
@@ -12106,6 +12124,9 @@ bool game::plmove(int dx, int dy)
             u.lifetime_stats()->squares_walked++;
         }
 
+        // Recalc dest_loc after possible shift
+        dest_loc = tripoint( x, y, u.posz() );
```

Comment:
At this point forget about `dest_loc` and use `u.pos3()` instead (in the following trap related code). It's the same (just called `u.set_x(x)` above) and per definition one wants to trigger the trap the character is on (trap location == player location is a logically required condition for triggering the trap).

This also allows to make `dest_loc` const, which reduces the complexity of the function.